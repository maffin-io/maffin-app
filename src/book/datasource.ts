import initSqlJs from 'sql.js';
import { DataSource } from 'typeorm';
import { useSWRConfig } from 'swr';

import {
  Account,
  Book,
  Commodity,
  Price,
  Split,
  Transaction,
} from './entities';

export async function initDB(database: Uint8Array): Promise<DataSource> {
  await initSqlJs({
    // Required to load the wasm binary asynchronously. Of course, you can host it wherever you want
    locateFile: file => `https://sql.js.org/dist/${file}`,
  });

  let datasource: DataSource;

  if (database.length !== 0) {
    datasource = new DataSource({
      type: 'sqljs',
      synchronize: true,
      database,
      logging: false,
      entities: [Account, Book, Commodity, Price, Split, Transaction],
    });
    await datasource.initialize();
  } else {
    datasource = new DataSource({
      type: 'sqljs',
      synchronize: true,
      logging: false,
      entities: [Account, Book, Commodity, Price, Split, Transaction],
    });
    await datasource.initialize();
    await createEmptyBook();
  }

  return datasource;
}

// TODO: Need to mutate the accounts SWR key so Root is shown!
async function createEmptyBook() {
  const rootAccount = Account.create({
    guid: 'rootAccount',
    name: 'Root',
    type: 'ROOT',
  });
  await Account.upsert(rootAccount, ['guid']);

  await Book.upsert(
    {
      guid: 'maffinBook',
      fk_root: rootAccount.guid,
    },
    ['guid'],
  );
}
