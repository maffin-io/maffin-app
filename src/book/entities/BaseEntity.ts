import {
  BaseEntity as BE,
  BeforeInsert,
  BeforeUpdate,
  PrimaryColumn,
  SaveOptions,
} from 'typeorm';
import * as v from 'class-validator';
import { QueryClient } from '@tanstack/react-query';

/**
 * Override BaseEntity so all entities have a generated guid in the format
 * we want and adds validation before saving or updating entities
 */
export default class BaseEntity extends BE {
  static CACHE_KEY = '';

  @BeforeInsert()
  @BeforeUpdate()
  async validate() {
    const errors = await v.validate(this, { stopAtFirstError: true });
    if (errors.length) {
      throw new Error(`validation failed: ${errors}`);
    }
  }

  @PrimaryColumn({
    type: 'varchar',
    length: 32,
  })
  @v.IsString()
    guid: string = crypto.randomUUID().substring(0, 31);

  async save(options?: SaveOptions): Promise<this> {
    const entity = await super.save(options);

    if (this.queryClient && ['Account', 'Commodity'].includes(this.constructor.name)) {
      updateCache({ queryClient: this.queryClient, entity });
    }

    return entity;
  }

  async remove(options?: SaveOptions): Promise<this> {
    const entity = await super.remove(options);

    if (this.queryClient && ['Account', 'Commodity'].includes(this.constructor.name)) {
      updateCache({ queryClient: this.queryClient, entity, isDelete: true });
    }

    return entity;
  }

  get queryClient(): QueryClient | undefined {
    // @ts-ignore
    return this.constructor.dataSource.options.extra?.queryClient;
  }
}

/**
 * Updates `/api/<entity_name>` and
 * `/api/<entity_name>/<guid>` keys so changes are propagated
 * in the cache.
 *
 * - /api/<entity_name> is updated as a list of Entity[]
 * - /api/<entity_name>/<guid> is updated with the value of the Entity
 */
export async function updateCache(
  {
    queryClient,
    entity,
    isDelete = false,
  }: {
    queryClient: QueryClient,
    entity: BaseEntity,
    isDelete?: boolean,
  },
) {
  // @ts-ignore
  const key = entity.constructor.CACHE_KEY;

  queryClient.setQueryData(
    [key],
    (entities: BaseEntity[] | undefined) => {
      if (!entities) {
        return undefined;
      }

      const newEntities = [...entities];

      const index = entities.findIndex(e => e.guid === entity.guid);
      if (index === -1) { // New entity added
        newEntities.push(entity);
      } else if (!isDelete) { // Entity updated
        newEntities[index] = entity;
      } else if (isDelete) { // Entity deleted
        newEntities.splice(index, 1);
      }

      return newEntities;
    },
  );

  queryClient.setQueryData(
    [key, { guid: entity.guid }],
    !isDelete ? entity : null,
  );
}
