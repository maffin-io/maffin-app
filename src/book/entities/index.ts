export { default as Account } from './Account';
export { default as BankConfig } from './BankConfig';
export { default as Commodity } from './Commodity';
export { default as Price } from './Price';
export { default as Split } from './Split';
export { default as Transaction } from './Transaction';
export { default as Book } from './Book';
export { default as BaseEntity } from './BaseEntity';
