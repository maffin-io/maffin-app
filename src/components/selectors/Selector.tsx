import React from 'react';
import { useHotkeys } from 'react-hotkeys-hook';
import Select, { SingleValue, components } from 'react-select';
import { BiSearch } from 'react-icons/bi';

export type SelectorProps<T> = {
  labelAttribute: string,
  options: T[],
  id?: string,
  placeholder?: string,
  onChange?: Function,
  isClearable?: boolean,
  defaultValue?: T,
  className?: string,
  disabled?: boolean,
};

export default function Selector<T extends object = {}>(
  {
    labelAttribute,
    options,
    placeholder,
    onChange,
    defaultValue,
    id = 'selector',
    isClearable = true,
    className = '',
    disabled = false,
  }: SelectorProps<T>,
): JSX.Element {
  const ref = React.useRef<HTMLDivElement>();
  useHotkeys('meta+k', () => {
    ref.current?.focus();
  });

  return (
    <Select
      // @ts-ignore
      ref={ref}
      id={id}
      name={id}
      aria-label={id}
      options={options}
      placeholder={placeholder || 'Choose an option'}
      onChange={(newValue: SingleValue<T> | null) => {
        if (onChange) {
          onChange(newValue);
        }
        ref.current?.blur();
      }}
      isClearable={isClearable}
      isDisabled={disabled}
      defaultValue={defaultValue}
      backspaceRemovesValue
      onKeyDown={(e: React.KeyboardEvent) => {
        if (e.key === 'Escape') {
          ref.current?.blur();
        }
      }}
      // @ts-ignore dunno how to set proper types here
      getOptionLabel={(option: T) => option[labelAttribute]}
      // @ts-ignore dunno how to set proper types here
      getOptionValue={(option: T) => option[labelAttribute]}
      openMenuOnFocus
      components={{ Control }}
      className={className}
      styles={{
        option: () => ({}),
      }}
      classNames={{
        // We set h-9 so it's the same as other inputs
        control: (state) => {
          const styles = '!h-9 !min-h-fit text-sm pl-10 !bg-gunmetal-800 !border-none';
          if (state.isDisabled) {
            // For some reason putting focus:bg-gunmetal-700 doesnt work
            return `opacity-50 ${styles}`;
          }
          return styles;
        },
        indicatorSeparator: () => 'hidden',
        menu: () => 'text-sm !bg-gunmetal-800',
        option: (state) => {
          const styles = 'w-full cursor-default hover:bg-gunmetal-700 px-4 py-2';
          if (state.isFocused) {
            // For some reason putting focus:bg-gunmetal-700 doesnt work
            return `bg-gunmetal-700 ${styles}`;
          }
          return styles;
        },
        singleValue: () => '!text-inherit',
        input: () => '!text-inherit',
      }}
    />
  );
}

function Control(
  {
    children,
    ...props
  }: any,
): JSX.Element {
  return (
    <components.Control {...props}>
      <BiSearch className="absolute text-md left-[10px]" />
      {children}
    </components.Control>
  );
}
