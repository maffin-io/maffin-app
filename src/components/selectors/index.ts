export { default as AccountSelector } from './AccountSelector';
export { default as CommoditySelector } from './CommoditySelector';
export { default as AccountTypeSelector } from './AccountTypeSelector';
export { default as NamespaceSelector } from './NamespaceSelector';

export type { AccountSelectorProps } from './AccountSelector';
export type { CommoditySelectorProps } from './CommoditySelector';
export type { AccountTypeSelectorProps } from './AccountTypeSelector';
