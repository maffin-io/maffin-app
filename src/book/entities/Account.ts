import * as v from 'class-validator';
import {
  Column,
  Entity, JoinColumn,
  ManyToOne,
  OneToMany,
  RelationId,
  Tree,
  TreeParent,
  TreeChildren,
  SaveOptions,
} from 'typeorm';
import type { QueryClient } from '@tanstack/react-query';

import {
  getAllowedSubAccounts,
  ASSET_ACCOUNTS,
  LIABILITY_ACCOUNTS,
} from '@/book/helpers/accountType';
import type Commodity from './Commodity';
import Split from './Split';
import BaseEntity from './BaseEntity';

/**
 * CREATE TABLE accounts (
 *   guid            CHAR(32) PRIMARY KEY NOT NULL,
 *   name            text(2048) NOT NULL,
 *   account_type    text(2048) NOT NULL,
 *   commodity_guid  CHAR(32) NOT NULL,
 *   commodity_scu   integer NOT NULL,
 *   non_std_scu     integer NOT NULL,
 *   parent_guid     CHAR(32),
 *   code            text(2048),
 *   description     text(2048),
 *   hidden          integer NOT NULL,
 *   placeholder     integer NOT NULL
 * );
 */

@Entity('accounts')
@Tree('nested-set')
export default class Account extends BaseEntity {
  static CACHE_KEY = ['api', 'accounts'];
  static TYPES = [
    'ROOT',
    'EQUITY',
    'INCOME',
    'EXPENSE',
    ...ASSET_ACCOUNTS,
    ...LIABILITY_ACCOUNTS,
  ];

  @Column({
    type: 'text',
    length: 2048,
  })
  @v.Matches(/[a-zA-Z.]+/)
  @v.Length(1, 2048)
    name!: string;

  @Column({
    type: 'text',
    enum: Account.TYPES,
    length: 2048,
    name: 'account_type',
  })
  @CheckAccountType()
  @v.IsIn(Account.TYPES)
    type!: string;

  path!: string;

  @TreeParent()
  @JoinColumn({ name: 'parent_guid' })
  @v.IsNotEmpty({ message: 'parent is required' })
  @v.ValidateIf(o => o.type !== 'ROOT')
  // Seems TreeParent doesn't let us create referencing the guid of the account
  // and needs the whole object...
    parent!: Account;

  @RelationId((account: Account) => account.parent)
    parentId: string;

  @TreeChildren()
    children!: Account[];

  @RelationId((account: Account) => account.children)
    childrenIds: string[];

  @ManyToOne('Commodity', { eager: true, cascade: true })
  @JoinColumn({ name: 'commodity_guid' })
  @CheckIECommodity()
  @v.ValidateIf(o => o.type !== 'ROOT')
  @v.IsNotEmpty({ message: 'commodity is required' })
    fk_commodity!: Commodity | string;

  get commodity(): Commodity {
    return this.fk_commodity as Commodity;
  }

  @OneToMany('Split', (split: Split) => split.fk_account)
    splits!: Split[];

  @Column({
    type: 'text',
    length: 2048,
    nullable: true,
  })
  @v.IsOptional()
  @v.Length(4, 2048)
    description?: string;

  @Column({
    default: false,
  })
    hidden!: boolean;

  @CheckPlaceholder()
  @Column({
    default: false,
  })
    // Placeholders are hierarchical accounts that don't containg
    // transactions, they are just parents of other accounts
    placeholder!: boolean;

  async save(options?: SaveOptions): Promise<this> {
    const account = await super.save(options);

    if (this.queryClient) {
      updateCache({ queryClient: this.queryClient });
    }

    return account;
  }

  async remove(options?: SaveOptions): Promise<this> {
    if (this.queryClient) {
      updateCache({ queryClient: this.queryClient });
    }

    const account = await super.remove(options);
    return account;
  }
}

export async function updateCache({ queryClient }: { queryClient: QueryClient }) {
  queryClient.invalidateQueries({
    queryKey: [...Account.CACHE_KEY],
  });
}

// https://github.com/typeorm/typeorm/issues/4714
Object.defineProperty(Account, 'name', { value: 'Account' });

/**
 * Checks allowed types for parent account given the current account's type
 */
function CheckAccountType(validationOptions?: v.ValidationOptions) {
  return function f(object: Account, propertyName: string) {
    v.registerDecorator({
      name: 'checkAccountType',
      target: object.constructor,
      propertyName,
      options: validationOptions,
      validator: {
        validate(type: string, args: v.ValidationArguments) {
          const account = args.object as Account;
          if (account.parent) {
            return getAllowedSubAccounts(account.parent.type).includes(type);
          }

          return true;
        },

        defaultMessage(args: v.ValidationArguments) {
          const account = args.object as Account;
          const allowedTypes = getAllowedSubAccounts(account.parent.type);

          return `only ${allowedTypes} types can be selected with ${account.parent.type} account as parent`;
        },
      },
    });
  };
}

/**
 * Checks that income/expense commodity is the same as the parent.
 *
 * This is a way to check that we always keep the same currency for income/expense
 * account as we only support them in the main currency.
 */
function CheckIECommodity(validationOptions?: v.ValidationOptions) {
  return function f(object: Account, propertyName: string) {
    v.registerDecorator({
      name: 'checkIECommodity',
      target: object.constructor,
      propertyName,
      options: validationOptions,
      validator: {
        validate(fk_commodity: Commodity, args: v.ValidationArguments) {
          const account = args.object as Account;
          if (
            account.parent
            && account.parent.type !== 'ROOT'
            && ['INCOME', 'EXPENSE'].includes(account.type)
          ) {
            return fk_commodity.guid === (account.parent.fk_commodity as Commodity).guid;
          }

          return true;
        },

        defaultMessage(args: v.ValidationArguments) {
          const account = args.object as Account;
          return `Income and Expense accounts must have ${(account.parent.fk_commodity as Commodity).mnemonic} as their commodity`;
        },
      },
    });
  };
}

/**
 * Checks if an account can be a placeholder. Accounts can be Placeholders
 * if they don't have any transactions associated.
 */
function CheckPlaceholder(validationOptions?: v.ValidationOptions) {
  return function f(object: Account, propertyName: string) {
    v.registerDecorator({
      name: 'checkPlaceholder',
      target: object.constructor,
      propertyName,
      options: validationOptions,
      validator: {
        async validate(_: string, args: v.ValidationArguments) {
          const account = args.object as Account;
          if (account.guid) {
            const [, numSplits] = await Split.findAndCount({
              where: { fk_account: { guid: account.guid } },
            });

            if (numSplits > 0 && account.placeholder) {
              return false;
            }
          }

          return true;
        },

        defaultMessage() {
          return 'Placeholder accounts cannot have transactions';
        },
      },
    });
  };
}
