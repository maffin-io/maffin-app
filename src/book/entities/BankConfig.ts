import {
  Column,
  Entity,
  OneToMany,
} from 'typeorm';
import {
  IsString,
} from 'class-validator';

import BaseEntity from './BaseEntity';
import type Account from './Account';

@Entity('bank_config')
export default class BankConfig extends BaseEntity {
  @Column({
    type: 'text',
    length: 2048,
  })
  @IsString()
    token!: string;

  @OneToMany(
    'Account',
    (account: Account) => account.config,
  )
    accounts!: Account[];
}

// https://github.com/typeorm/typeorm/issues/4714
Object.defineProperty(BankConfig, 'name', { value: 'BankConfig' });
