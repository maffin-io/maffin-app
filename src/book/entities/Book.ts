import {
  Entity,
  JoinColumn,
  OneToOne,
} from 'typeorm';

import BaseEntity from './BaseEntity';
import Account from './Account';

/**
 * https://wiki.gnucash.org/wiki/SQL#Books
 * CREATE TABLE books (
 *   guid                CHAR(32) PRIMARY KEY NOT NULL,
 *   root_account_guid   CHAR(32) NOT NULL,
 *   root_template_guid  CHAR(32) NOT NULL
 * );
*/

@Entity('books')
export default class Book extends BaseEntity {
  @OneToOne(() => Account, { eager: true })
  @JoinColumn({ name: 'root_account_guid' })
    fk_root!: Account | string;

  get root(): Account {
    return this.fk_root as Account;
  }
}

// https://github.com/typeorm/typeorm/issues/4714
Object.defineProperty(Book, 'name', { value: 'Book' });
