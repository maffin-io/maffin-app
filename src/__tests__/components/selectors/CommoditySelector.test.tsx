import React from 'react';
import { render, screen } from '@testing-library/react';
import { SWRConfig } from 'swr';

import Selector from '@/components/selectors/Selector';
import { Commodity } from '@/book/entities';
import { CommoditySelector } from '@/components/selectors';

jest.mock('@/components/selectors/Selector', () => jest.fn(
  () => <div data-testid="Selector" />,
));

describe('CommoditySelector', () => {
  beforeEach(() => {
    jest.spyOn(Commodity, 'find').mockResolvedValue([]);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders as expected', async () => {
    const { container } = render(
      <SWRConfig value={{ provider: () => new Map() }}>
        <CommoditySelector />
      </SWRConfig>,
    );

    await screen.findByTestId('Selector');
    expect(Selector).toHaveBeenCalledWith(
      {
        id: 'commoditySelector',
        isClearable: true,
        defaultValue: undefined,
        className: '',
        labelAttribute: 'mnemonic',
        options: [],
        placeholder: 'Choose commodity',
        onChange: expect.any(Function),
      },
      {},
    );
    expect(container).toMatchSnapshot();
  });

  it('passes data as expected', async () => {
    const mockOnSave = jest.fn();
    const { container } = render(
      <SWRConfig value={{ provider: () => new Map() }}>
        <CommoditySelector
          id="customId"
          placeholder="My placeholder"
          isClearable={false}
          className="class"
          defaultValue={{ mnemonic: 'EUR' } as Commodity}
          onChange={mockOnSave}
        />
      </SWRConfig>,
    );

    await screen.findByTestId('Selector');
    expect(Selector).toHaveBeenCalledWith(
      {
        id: 'customId',
        isClearable: false,
        defaultValue: { mnemonic: 'EUR' },
        className: 'class',
        labelAttribute: 'mnemonic',
        options: [],
        placeholder: 'My placeholder',
        onChange: mockOnSave,
      },
      {},
    );
    expect(container).toMatchSnapshot();
  });

  it('loads commodities and passes as options', async () => {
    const options = [
      {
        mnemonic: 'EUR',
        namespace: 'CURRENCY',
      } as Commodity,
      {
        mnemonic: 'USD',
        namespace: 'CURRENCY',
      } as Commodity,
      {
        mnemonic: 'IDVY.AS',
        namespace: 'AS',
      } as Commodity,
    ];
    jest.spyOn(Commodity, 'find').mockResolvedValue(options);

    render(
      <SWRConfig value={{ provider: () => new Map() }}>
        <CommoditySelector />
      </SWRConfig>,
    );

    await screen.findByTestId('Selector');
    expect(Selector).toHaveBeenCalledWith(
      expect.objectContaining({
        options,
      }),
      {},
    );
  });

  it('filters commodities', async () => {
    const options = [
      {
        mnemonic: 'EUR',
        namespace: 'CURRENCY',
      } as Commodity,
      {
        mnemonic: 'USD',
        namespace: 'CURRENCY',
      } as Commodity,
      {
        mnemonic: 'IDVY.AS',
        namespace: 'AS',
      } as Commodity,
    ];
    jest.spyOn(Commodity, 'find').mockResolvedValue(options);

    render(
      <SWRConfig value={{ provider: () => new Map() }}>
        <CommoditySelector
          ignoreNamespaces={['CURRENCY']}
        />
      </SWRConfig>,
    );

    await screen.findByTestId('Selector');
    expect(Selector).toHaveBeenCalledWith(
      expect.objectContaining({
        options: [options[2]],
      }),
      {},
    );
  });

  it('gets commodities once only', async () => {
    const { rerender } = render(
      <CommoditySelector />,
    );

    rerender(
      <CommoditySelector />,
    );

    await screen.findByTestId('Selector');
    expect(Commodity.find).toHaveBeenCalledTimes(1);
  });
});
