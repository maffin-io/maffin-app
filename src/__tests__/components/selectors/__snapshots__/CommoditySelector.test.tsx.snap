// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommoditySelector passes data as expected 1`] = `
<div>
  <div
    data-testid="Selector"
  />
</div>
`;

exports[`CommoditySelector renders as expected 1`] = `
<div>
  <div
    data-testid="Selector"
  />
</div>
`;
